### ISSUE TITLE
Cannot mount managed disk on Standard_DC4s/DC2s instances

### ISSUE THREAD

**Issue by joekottke** [kubernetes-sigs/azuredisk-csi-driver#212]
**What happened**:
I'm trying to mount a managed disk into a Pod running on a `Standard_DC4s` instance, via PVC and automatic provisioning. The PVC and PV are created, the disk is provisioned, and the disk is attached to the machine (I can see it as `/dev/sdc` and can even get output from `sudo fdisk -l /dev/sdc`, but the the Pod never initializes, and results in this error:

```
  Type     Reason                  Age               From                                       Message
  ----     ------                  ----              ----                                       -------
  Normal   Scheduled               110s              default-scheduler                          Successfully assigned joekottke/node1-55fccb9cf-7cwm8 to k8s-sgxpool1-33251224-vmss000003
  Normal   SuccessfulAttachVolume  68s               attachdetach-controller                    AttachVolume.Attach succeeded for volume "pvc-cc5712cb-6e0a-4ffa-89be-b2cce8a1147b"
  Warning  FailedMount             8s (x6 over 41s)  kubelet, k8s-sgxpool1-33251224-vmss000003  MountVolume.MountDevice failed for volume "pvc-cc5712cb-6e0a-4ffa-89be-b2cce8a1147b" : rpc error: code = Internal desc = Failed to find disk and lun 0. azureDisk - findDiskByLun(0) failed within timeout
```
I am using node labels and nodeSelectors to get pods to run on these specific machines.  When I schedule the pod on a `Standard_D4_v3` or `Standard_D2_v3` instance, it works just fine.

I increased the logging on the `azuredisk` container to capture the log output.

On the `Standard_D4_v3` machine I see this:

```
I1206 18:43:04.835554       1 azure_common_linux.go:45] Azure sys disks paths: [sdb sdb1 sda sda1 sda14 sda15]
I1206 18:43:04.835948       1 azure_common_linux.go:157] azureDisk - validating disk "sda" with sys disk "sdb"
I1206 18:43:04.835976       1 azure_common_linux.go:157] azureDisk - validating disk "sda" with sys disk "sdb1"
I1206 18:43:04.835989       1 azure_common_linux.go:157] azureDisk - validating disk "sda" with sys disk "sda"
I1206 18:43:04.836187       1 azure_common_linux.go:157] azureDisk - validating disk "sdb" with sys disk "sdb"
I1206 18:43:04.836348       1 azure_common_linux.go:157] azureDisk - validating disk "sdc" with sys disk "sdb"
I1206 18:43:04.836359       1 azure_common_linux.go:157] azureDisk - validating disk "sdc" with sys disk "sdb1"
I1206 18:43:04.836366       1 azure_common_linux.go:157] azureDisk - validating disk "sdc" with sys disk "sda"
I1206 18:43:04.836373       1 azure_common_linux.go:157] azureDisk - validating disk "sdc" with sys disk "sda1"
I1206 18:43:04.836379       1 azure_common_linux.go:157] azureDisk - validating disk "sdc" with sys disk "sda14"
I1206 18:43:04.836386       1 azure_common_linux.go:157] azureDisk - validating disk "sdc" with sys disk "sda15"
I1206 18:43:04.836438       1 azure_common_linux.go:52] azureDisk - begin to find sdc from /devhost/disk/azure/scsi1/
I1206 18:43:04.836445       1 azure_common_linux.go:56] azureDisk - begin to Readlink: /devhost/disk/azure/scsi1/lun0
```

The contents of `/dev/disk/azure` on that machine is as follows:

```
/dev/disk/azure/
├── resource -> ../../sdb
├── resource-part1 -> ../../sdb1
├── root -> ../../sda
├── root-part1 -> ../../sda1
├── root-part14 -> ../../sda14
├── root-part15 -> ../../sda15
└── scsi1
    └── lun0 -> ../../../sdc
```

On the `Standard_DC4s` machine I see this in the logs:

```
I1206 18:38:44.745470       1 azure_common_linux.go:45] Azure sys disks paths: []
I1206 18:38:44.745535       1 azure_common_linux.go:104] /dev/disk/azure is not populated, now try to parse 0:0:0:0 directly
I1206 18:38:44.745542       1 azure_common_linux.go:104] /dev/disk/azure is not populated, now try to parse 0:0:0:1 directly
I1206 18:38:44.745545       1 azure_common_linux.go:104] /dev/disk/azure is not populated, now try to parse 1:0:0:0 directly
E1206 18:38:44.745557       1 utils.go:117] GRPC error: rpc error: code = Internal desc = Failed to find disk and lun 0. azureDisk - findDiskByLun(0) failed within timeout
```

And the contents of `/dev/disk/azure` look different:

```
/dev/disk/azure
├── scsi0
│   ├── lun0 -> ../../../sda
│   ├── lun0-part1 -> ../../../sda1
│   ├── lun0-part14 -> ../../../sda14
│   ├── lun0-part15 -> ../../../sda15
│   ├── lun1 -> ../../../sdb
│   └── lun1-part1 -> ../../../sdb1
└── scsi1
    └── lun0 -> ../../../sdc
```

I'm not sure why this is happening.  Is it because the `Standard_DC4s` instances are using managed disks under the hood for their OS Disk?  I don't know.

**What you expected to happen**:

Managed disks should be able to attach correctly to a pod scheduled on a `Standard_DC4s` 
node instance.

**How to reproduce it**:

1. Run an agentpool that is comprised of `Standard_DC2s` or `Standard_DC4s` instances.
2. Label those nodes so that they can be used specifically for scheduling. Example uses the `sgx-enabled-node: "true"` label
3. Install `azuredisk-csi-driver` from master
4. Schedule a pod with the following Yaml:

```
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: node1-data
spec:
  storageClassName: disk.csi.azure.com
  accessModes:
    - "ReadWriteOnce"
  resources:
    requests:
      storage: 10Gi
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: node1
spec:
  replicas: 1
  strategy:
    type: Recreate
  selector:
    matchLabels:
      nodename: node1
  template:
    metadata:
      labels:
        nodename: node1
    spec:
      nodeSelector:
        sgx-enabled-node: "true"
      containers:
      - image: nginx
        name: nginx-azuredisk
        command:
        - "/bin/sh"
        - "-c"
        - while true; do echo $(date) >> /mnt/azuredisk/outfile; sleep 1; done
        volumeMounts:
        - name: azuredisk01
          mountPath: "/mnt/azuredisk"
  volumes:
  - name: azuredisk01
    persistentVolumeClaim:
      claimName: node1-data
```

**Anything else we need to know?**:

**Environment**:
- CSI Driver version: `v0.5.0` (pulled from master today)
- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"15", GitVersion:"v1.15.3", GitCommit:"2d3c76f9091b6bec110a5e63777c332469e0cba2", GitTreeState:"clean", BuildDate:"2019-08-19T12:36:28Z", GoVersion:"go1.12.9", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"15", GitVersion:"v1.15.3", GitCommit:"2d3c76f9091b6bec110a5e63777c332469e0cba2", GitTreeState:"clean", BuildDate:"2019-08-19T11:05:50Z", GoVersion:"go1.12.9", Compiler:"gc", Platform:"linux/amd64"}
```
- OS (e.g. from /etc/os-release):
```
NAME="Ubuntu"
VERSION="18.04.3 LTS (Bionic Beaver)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 18.04.3 LTS"
VERSION_ID="18.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=bionic
UBUNTU_CODENAME=bionic
```
- Kernel (e.g. `uname -a`): `4.18.0-1025-azure #27~18.04.1-Ubuntu SMP Thu Jul 4 20:49:16 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux`
- Install tools: `aks-engine v0.39.2`
- Others:

**Comment by joekottke** [comment]
Also discovered this with the in-tree `azureDisk` driver.  I was able to diagnose better with the CSI driver, but I bet this issue exists in both codebases.

**Comment by joekottke** [comment]
One more thing.  It looks like the disk gets attached at `1:0:0:0` in `/sys/bus/scsi/devices`:

```
/sys/bus/scsi/devices/
├── 0:0:0:0 -> ../../../devices/LNXSYSTM:00/LNXSYBUS:00/ACPI0004:00/VMBUS:00/f8b3781a-1e82-4818-a1c3-63d806ec15bb/host0/target0:0:0/0:0:0:0
├── 0:0:0:1 -> ../../../devices/LNXSYSTM:00/LNXSYBUS:00/ACPI0004:00/VMBUS:00/f8b3781a-1e82-4818-a1c3-63d806ec15bb/host0/target0:0:0/0:0:0:1
├── 1:0:0:0 -> ../../../devices/LNXSYSTM:00/LNXSYBUS:00/ACPI0004:00/VMBUS:00/f8b3781b-1e82-4818-a1c3-63d806ec15bb/host1/target1:0:0/1:0:0:0
├── host0 -> ../../../devices/LNXSYSTM:00/LNXSYBUS:00/ACPI0004:00/VMBUS:00/f8b3781a-1e82-4818-a1c3-63d806ec15bb/host0
├── host1 -> ../../../devices/LNXSYSTM:00/LNXSYBUS:00/ACPI0004:00/VMBUS:00/f8b3781b-1e82-4818-a1c3-63d806ec15bb/host1
├── target0:0:0 -> ../../../devices/LNXSYSTM:00/LNXSYBUS:00/ACPI0004:00/VMBUS:00/f8b3781a-1e82-4818-a1c3-63d806ec15bb/host0/target0:0:0
└── target1:0:0 -> ../../../devices/LNXSYSTM:00/LNXSYBUS:00/ACPI0004:00/VMBUS:00/f8b3781b-1e82-4818-a1c3-63d806ec15bb/host1/target1:0:0
```
```
/dev/disk/azure
├── scsi0
│   ├── lun0 -> ../../../sda
│   ├── lun0-part1 -> ../../../sda1
│   ├── lun0-part14 -> ../../../sda14
│   ├── lun0-part15 -> ../../../sda15
│   ├── lun1 -> ../../../sdb
│   └── lun1-part1 -> ../../../sdb1
└── scsi1
    └── lun0 -> ../../../sdc
```
And the code in `azure_common_linux.go` assumes that when `target <= 3` is it most likely a system/os disk.  So that seems to be my issue (I think)

**Comment by joekottke** [comment]
I have a _very_ rudimentary workaround (that probably only just works for me, and probably is destructive elsewhere) [in this branch](https://github.com/joekottke/azuredisk-csi-driver/tree/standard_dc4s_workaround)

**Comment by andyzhangx** [comment]
> I have a _very_ rudimentary workaround (that probably only just works for me, and probably is destructive elsewhere) [in this branch](https://github.com/joekottke/azuredisk-csi-driver/tree/standard_dc4s_workaround)

@joekottke thanks for reporting this issue, it's quite detailed and obvious that current in-tree and csi azure disk driver does not work on Standard_DC series VM, I will make a PR per your instructions, thanks!

**Comment by joekottke** [comment]
@andyzhangx Have you made any progress on this issue?  My fix works for me, so I'm not in any great hurry, but I would rather have a fix that is part of the main source.

**Comment by andyzhangx** [comment]
> @andyzhangx Have you made any progress on this issue? My fix works for me, so I'm not in any great hurry, but I would rather have a fix that is part of the main source.

@joekottke already fixed this issue on master branch now, you could try the image: `mcr.microsoft.com/k8s/csi/azuredisk-csi:v0.5.0`

**Comment by andyzhangx** [comment]
Update: this issue is also fixed in 1.15.8, 1.16.5, 1.17.1, 1.18.0 for in-tree azure disk driver.
